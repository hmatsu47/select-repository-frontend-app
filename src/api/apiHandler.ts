import { fetchWithTimeout } from "./fetchWithTimeout";

// ベース URI
export const baseUri = "http://10.11.17.43/api";

// APIからデータ取得（GET）
export const getApiData = async (url: string) => {
  return await (await fetchWithTimeout(url)).json();
};
// APIからデータ取得（POST）
export const postApiData = async (url: string, object: Object) => {
  return await (await fetchWithTimeout(url, object)).json();
};
