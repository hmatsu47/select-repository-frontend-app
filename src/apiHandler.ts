import { fetchWithTimeout } from "./fetchWithTimeout";

// APIからデータ取得（GET）
export const getApiData = async (url: string) => {
  return await (await fetchWithTimeout(url)).json();
};
// APIからデータ取得（POST）
export const postApiData = async (url: string, object: Object) => {
  return await (await fetchWithTimeout(url, object)).json();
};
