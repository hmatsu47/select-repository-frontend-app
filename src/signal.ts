import { createSignal } from "solid-js";
import { AlertColor } from "@suid/material/Alert";

import { ImageItem, RepositoryItem, ServiceItem } from "./type";

export const [service, setService] = createSignal<string | undefined>();
export const [repository, setRepository] = createSignal<string | undefined>();
export const [services, setServices] = createSignal<
  ServiceItem[] | undefined
>();
export const [repositories, setRepositories] = createSignal<
  RepositoryItem[] | undefined
>();
export const [images, setImages] = createSignal<ImageItem[] | undefined>();
export const [imageUri, setImageUri] = createSignal<string | undefined>();
export const [releaseAt, setReleaseAt] = createSignal<Date | undefined>();
export const [lastImageUri, setLastImageUri] = createSignal<
  string | undefined
>();
export const [lastReleasedAt, setLastReleasedAt] = createSignal<
  Date | undefined
>();
export const [isReleased, setIsReleased] = createSignal<boolean | undefined>();
export const [message, setMessage] = createSignal<string | undefined>();
export const [messageSeverity, setMessageSeverity] = createSignal<
  AlertColor | undefined
>();
