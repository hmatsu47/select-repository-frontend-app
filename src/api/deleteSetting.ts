import { ErrorResponse } from "../type";
import { baseUri, deleteApiData } from "./apiHandler";
import {
  imageUri,
  isReleased,
  isReleaseSelected,
  releaseAt,
  service,
  setMessage,
  setMessageSeverity,
  setImageUri,
  setIsReleaseSelected,
  setReleaseAt,
} from "../signal";

export const deleteSetting = async () => {
  if (
    !isReleased() === undefined ||
    !isReleaseSelected() ||
    !imageUri() ||
    !releaseAt()
  ) {
    return;
  }
  const load = async (): Promise<void> => {
    const data: null | ErrorResponse = await deleteApiData(
      `${baseUri}/setting/${service()}`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setMessage((data as ErrorResponse).message);
      setMessageSeverity("error");
      return;
    }
    setImageUri(undefined);
    setReleaseAt(undefined);
    setIsReleaseSelected(false);
    setMessage("リリースを取り消しました。");
    setMessageSeverity("success");
  };
  void load();
};
