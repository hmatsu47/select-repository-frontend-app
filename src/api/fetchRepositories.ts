import { ErrorResponse, RepositoryItem } from "../type";
import { baseUri, getApiData } from "./apiHandler";
import {
  service,
  setImages,
  setMessage,
  setMessageSeverity,
  setRepositories,
} from "../signal";

export const fetchRepositories = async () => {
  if (!service()) {
    return;
  }
  const load = async (): Promise<void> => {
    const data: RepositoryItem[] | ErrorResponse = await getApiData(
      `${baseUri}/repositories/${service()}`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setRepositories(undefined);
      setImages(undefined);
      setMessage((data as ErrorResponse).message);
      setMessageSeverity("error");
      return;
    }
    setRepositories(data as RepositoryItem[]);
    setImages(undefined);
    setMessage(undefined);
  };
  void load();
};
