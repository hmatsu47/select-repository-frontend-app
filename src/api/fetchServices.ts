import { ErrorResponse, ServiceItem } from "../type";
import { baseUri, getApiData } from "./apiHandler";
import {
  setErrorMessage,
  setImages,
  setRepositories,
  setServices,
} from "../signal";

export const fetchServices = async () => {
  const load = async (): Promise<void> => {
    const data: ServiceItem[] | ErrorResponse = await getApiData(
      `${baseUri}/services`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setServices(undefined);
      setRepositories(undefined);
      setImages(undefined);
      setErrorMessage((data as ErrorResponse).message);
      return;
    }
    setServices(data as ServiceItem[]);
    setRepositories(undefined);
    setImages(undefined);
    setErrorMessage(undefined);
  };
  void load();
};
