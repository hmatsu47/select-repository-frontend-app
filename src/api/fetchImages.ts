import { ErrorResponse, ImageItem } from "../type";
import { baseUri, getApiData } from "./apiHandler";
import { repository, service, setErrorMessage, setImages } from "../signal";

export const fetchImages = async () => {
  if (!service() || !repository()) {
    return;
  }
  const load = async (): Promise<void> => {
    const data: ImageItem[] | ErrorResponse = await getApiData(
      `${baseUri}/images/${service()}/${repository()}`
    );
    if (
      typeof data === "object" &&
      data !== null &&
      typeof (data as ErrorResponse).message === "string"
    ) {
      // 戻り値がエラーメッセージの場合
      setImages(undefined);
      setErrorMessage((data as ErrorResponse).message);
      return;
    }
    setImages(data as ImageItem[]);
    setErrorMessage(undefined);
  };
  void load();
};
