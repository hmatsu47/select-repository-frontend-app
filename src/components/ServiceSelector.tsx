import { For, onMount, Show } from "solid-js";
import Box from "@suid/material/Box";
import Stack from "@suid/material/Stack";
import Typography from "@suid/material/Typography";
import { ServiceButton } from "./ServiceButton";
import { fetchServices } from "../api/fetchServices";
import { services, setService } from "../signal";

export const ServiceSelector = () => {
  onMount(async () => {
    await fetchServices();
    // ローカルストレージから前回の選択肢を読み取る
    const tmpService = localStorage.getItem("selectedService");
    if (tmpService) {
      setService(tmpService);
    }
  });

  return (
    <Box
      sx={{
        width: "100%",
        minWidth: "1024px",
        display: "flex",
      }}
      aria-live="polite"
    >
      <Show when={services()} fallback={<div>Loading...</div>}>
        <Stack spacing={2} direction="row">
          <Typography variant="h6" sx={{ minWidth: "256px" }}>
            リリース対象サービス
          </Typography>
          <For each={services()} fallback={<></>}>
            {(serviceItem) => <ServiceButton serviceItem={serviceItem} />}
          </For>
        </Stack>
      </Show>
    </Box>
  );
};
