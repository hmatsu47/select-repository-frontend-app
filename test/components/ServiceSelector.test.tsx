// @vitest-environment jsdom

import { describe, expect, test } from "vitest";
import "vi-fetch/setup";
import { mockGet } from "vi-fetch";
import { render } from "solid-testing-library";
import { formatSnapshot } from "../common/formatSnapshot";
import { ServiceItem } from "../../src/type";
import { baseUri } from "../../src/api/apiHandler";
import { ServiceSelector } from "../../src/components/ServiceSelector";
import { setServices } from "../../src/signal";

describe("<ServiceSelector />", () => {
  const serviceSelectorList = [
    {
      title: "サービス x1",
      services: [{ name: "service1" } as ServiceItem],
    },
    {
      title: "サービス x2",
      services: [
        { name: "service1" } as ServiceItem,
        { name: "service2" } as ServiceItem,
      ],
    },
    {
      title: "サービス x3",
      services: [
        { name: "service1" } as ServiceItem,
        { name: "service2" } as ServiceItem,
        { name: "service3" } as ServiceItem,
      ],
    },
  ];
  beforeEach(() => {
    // localStorage はモック化すべきかも（うまくいかなかったため一旦モックなしで実装）
    localStorage.removeItem("selectedService");
  });
  serviceSelectorList.forEach((testCase) => {
    test(testCase.title, async () => {
      setServices(testCase.services);
      const mockServices = mockGet(`${baseUri}/services`).willResolve(
        testCase.services
      );
      const { container, findByText, unmount } = render(() => (
        <ServiceSelector />
      ));
      expect(mockServices).toHaveFetched();
      // css の名前が動的に変わるので固定値に置換
      const html = formatSnapshot(container.innerHTML);
      expect(html).toMatchSnapshot();
      // 各ボタン
      testCase.services.forEach(async (service) => {
        const button = (await findByText(service.name)) as HTMLElement;
        expect(button).toHaveTextContent(service.name);
      });
      unmount();
      mockServices.clear();
    });
  });
});
