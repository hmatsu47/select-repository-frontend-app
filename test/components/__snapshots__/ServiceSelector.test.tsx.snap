// Vitest Snapshot v1

exports[`<ServiceSelector /> > サービス x1 1`] = `"<div aria-live=\\"polite\\" class=\\"MuiBox-root css-xxxxxxxx\\"><div class=\\"MuiStack-root css-xxxxxxxx\\"><h6 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-root css-xxxxxxxx\\">リリース対象サービス</h6><button type=\\"button\\" title=\\"service1 を選択\\" tabindex=\\"0\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxxxx\\">service1<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button></div></div>"`;

exports[`<ServiceSelector /> > サービス x2 1`] = `"<div aria-live=\\"polite\\" class=\\"MuiBox-root css-xxxxxxxx\\"><div class=\\"MuiStack-root css-xxxxxxxx\\"><h6 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-root css-xxxxxxxx\\">リリース対象サービス</h6><button type=\\"button\\" title=\\"service1 を選択\\" tabindex=\\"0\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxxxx\\">service1<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button><button type=\\"button\\" title=\\"service2 を選択\\" tabindex=\\"0\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxxxx\\">service2<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button></div></div>"`;

exports[`<ServiceSelector /> > サービス x3 1`] = `"<div aria-live=\\"polite\\" class=\\"MuiBox-root css-xxxxxxxx\\"><div class=\\"MuiStack-root css-xxxxxxxx\\"><h6 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-root css-xxxxxxxx\\">リリース対象サービス</h6><button type=\\"button\\" title=\\"service1 を選択\\" tabindex=\\"0\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxxxx\\">service1<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button><button type=\\"button\\" title=\\"service2 を選択\\" tabindex=\\"0\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxxxx\\">service2<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button><button type=\\"button\\" title=\\"service3 を選択\\" tabindex=\\"0\\" class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorInherit MuiButton-root MuiButtonBase-root css-xxxxxxxx\\">service3<span class=\\"MuiTouchRipple-root css-xxxxxxxx\\"></span></button></div></div>"`;
